### ISSUE TITLE
maybe_idx allocating for heterogeneous systems

### ISSUE THREAD

**Issue by lindnemi** [PIK-ICoNe/NetworkDynamics.jl#85]
for heterogeneous systems, we are using Arrays of abstract type VertexFunction. In this case maybe_idx is allocating.

Because the f! of each Vertex may be different, we get a type instability :Any

Maybe we could take the inital vertexlist and keep only the unique elements, as well as a set of node indices for each element and then iterate over those

**Comment by hexaeder** [comment]
you mean that instead of
```julia
struct nd_ODE_Static{G, GD, T1, T2}
    vertices!::T1
    edges!::T2
end
```
where `vertices <: Array{<ODEVertex}` we would get something like 
```julia
vertices <: Tuple{ODEVertex{type1}, ODEVertex{type2}, ODEVertex{type3}, ...}
```
together with some indices 
```julia
vierticeidx <: NTuple{nr_off_diff_vertex_functions, Vector{Int}}
```
and than we'd have a separate homogenious core loop over every "type" of vertex function?

```julia
for (i, odevert) in enumerate(vertices)
   indices = verticeidx[i]
   # do core loop for odevert and indices
end
```
I'm pretty sure `vertices[i]` is not type stable unless `i` is a compile time constant. We have to try whether the type inference works through `for odevert in vertices`.... otherwise one might need to use a `@generated` function to manually unroll the loop over the N different vertex types...

**Comment by lindnemi** [comment]
i already checked it, and the tuple loop is not type stable :(

**Comment by lindnemi** [comment]
at the moment i manually unrolled the tuple (just testing with 2 different vertices), but good to know that @generated is how you would do such a thing normally.

**Comment by hexaeder** [comment]
Well it is one possible solution the information is there to make it typestable. We could also use a package like https://github.com/StephenVavasis/Unroll.jl or just write a littel `@unroll` macro ourselfs. I haven't thought this throug yet but if it works it is much more straight forward than using `@generated` functions

**Comment by hexaeder** [comment]
is it typstable with the manually unrolled loop? I am not 100% sure that `getindex(::Tuple, i)` is stable for compile time constant i

**Comment by lindnemi** [comment]
its type stable

**Comment by lindnemi** [comment]
there are still allocations though, i just have no idea why

**Comment by lindnemi** [comment]
partially solved with #91. ODE ODE and DDE Static cases remain open. Remove `maybe_idx` when this is finished.

**Comment by lindnemi** [comment]
https://github.com/PIK-ICoNe/NetworkDynamics.jl/pull/91#discussion_r687443366

**Comment by lindnemi** [comment]
solved by #98
